// jarvis-common/src/jarvis/action-date-filter.ts
import moment, { Moment } from 'moment';

/**
 * Constructs a date range filter query accounting for actions which span multiple days.
 * @param startDateKey - The key for the start date.
 * @param filterStart - The start date for the filter.
 * @param filterEnd - The end date for the filter.
 * @param endDateKey - An optional key for the end date.
 * @returns An object representing the date range filter query.
 */
export const getDateRangeFilterQuery = (
  startDateKey: string,
  filterStart: Date | Moment | string,
  filterEnd: Date | Moment | string,
  endDateKey = '',
) => {
  const momentStart = moment.isMoment(filterStart)
    ? filterStart
    : moment(filterStart);
  const momentEnd = moment.isMoment(filterEnd) ? filterEnd : moment(filterEnd);

  const isSameDay = momentStart.isSame(momentEnd, 'day');
  const rangeStart = momentStart.clone().utc().startOf('day');
  const rangeEnd = (isSameDay ? rangeStart.clone() : momentEnd.clone())
    .utc()
    .endOf('day');

  let filter: any = {
    [startDateKey]: {
      between: {
        lower: rangeStart,
        upper: rangeEnd,
      },
    },
  };

  if (endDateKey) {
    filter = {
      or: [
        {
          [startDateKey]: { between: { lower: rangeStart, upper: rangeEnd } },
        },
        {
          and: [
            { [startDateKey]: { lt: rangeStart } },
            { [endDateKey]: { gt: rangeEnd } },
          ],
        },
        {
          [endDateKey]: { between: { lower: rangeStart, upper: rangeEnd } },
        },
      ],
    };
  }
  return filter;
};
