// jarvis-common/dist/jarvis/action-date-filter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDateRangeFilterQuery = void 0;
const moment_1 = __importDefault(require("moment"));
const getDateRangeFilterQuery = (startDateKey, filterStart, filterEnd, endDateKey = '') => {
    const momentStart = moment_1.default.isMoment(filterStart)
        ? filterStart
        : (0, moment_1.default)(filterStart);
    const momentEnd = moment_1.default.isMoment(filterEnd) ? filterEnd : (0, moment_1.default)(filterEnd);
    const isSameDay = momentStart.isSame(momentEnd, 'day');
    const rangeStart = momentStart.clone().utc().startOf('day');
    const rangeEnd = (isSameDay ? rangeStart.clone() : momentEnd.clone())
        .utc()
        .endOf('day');
    let filter = {
        [startDateKey]: {
            between: {
                lower: rangeStart,
                upper: rangeEnd,
            },
        },
    };
    if (endDateKey) {
        filter = {
            or: [
                {
                    [startDateKey]: { between: { lower: rangeStart, upper: rangeEnd } },
                },
                {
                    and: [
                        { [startDateKey]: { lt: rangeStart } },
                        { [endDateKey]: { gt: rangeEnd } },
                    ],
                },
                {
                    [endDateKey]: { between: { lower: rangeStart, upper: rangeEnd } },
                },
            ],
        };
    }
    return filter;
};
exports.getDateRangeFilterQuery = getDateRangeFilterQuery;
