// jarvis-common/dist/jarvis/formatting-helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getActionTitle = getActionTitle;
exports.formatActionStatus = formatActionStatus;
const moment_1 = __importDefault(require("moment"));
function getActionTitle(action) {
    if (action?.title?.startsWith('-')) {
        return `${action?.date ? `${moment_1.default.utc(action.date).format('D MMMM')} - ` : ''}${(action?.location?.city ? action?.location?.city : '').concat(action.location?.province
            ? ', ' + action?.location?.province
            : `${action?.location?.country ? ', ' + action?.location?.country : ''}`)} : ${action?.status ? `(${formatActionStatus(action?.actionStatus)}) ` : ''}${action?.issues && action?.issues?.length > 0
            ? action?.issues?.map((x) => x?.name).join(', ')
            : ''} ${action?.tactic?.name ? action?.tactic?.name : ''}`;
    }
    else {
        return `${action?.date ? `${moment_1.default.utc(action?.date).format('D MMMM')} - ` : ''}${action?.title}`;
    }
}
var ActionStatus;
(function (ActionStatus) {
    ActionStatus["Updated"] = "Updated";
    ActionStatus["Rescheduled"] = "Rescheduled";
    ActionStatus["LocationChanged"] = "LocationChanged";
    ActionStatus["ImpactIncreased"] = "ImpactIncreased";
    ActionStatus["Canceled"] = "Canceled";
    ActionStatus["Reviewed"] = "Reviewed";
})(ActionStatus || (ActionStatus = {}));
function formatActionStatus(status) {
    if (!status)
        return 'Created';
    switch (status) {
        case ActionStatus.Canceled:
            return 'Canceled';
        case ActionStatus.Updated:
            return 'Updated';
        case ActionStatus.ImpactIncreased:
            return 'Risk Increased';
        case ActionStatus.LocationChanged:
            return 'Location Changed';
        case ActionStatus.Rescheduled:
            return 'Rescheduled';
        case ActionStatus.Reviewed:
            return 'Reviewed';
        default:
            return '';
    }
}
